### ISSUE TITLE
Add h:inputFile component

### ISSUE THREAD

**Issue by glassfishrobot** [javaee/javaserverfaces-spec#690]
Add an h:inputFile component for file upload.

Note that this may require Servlet 3.0 - which previously we'd avoided as a
requirement.
#### Environment
Operating System: All
Platform: All
#### Affected Versions
[2.0]

**Comment by glassfishrobot** [comment]
Reported by driscoll

**Comment by glassfishrobot** [comment]
Was assigned to rogerk

**Comment by glassfishrobot** [comment]
mojavelinux said:
In addition to the obvious binding to java.io.File, I'd like to see support for
binding to a byte[] property. This lowers the barrier for being able to put the
data into persistence storage.

Hopefully we can draw from Seam's s:fileUpload as a prototype. Seam allows
multiple bindings. One to the property that accepts the data and another to the
property that accepts the content type. This makes it adaptable to existing
entities. Just something to think about.

**Comment by glassfishrobot** [comment]
mwessendorf@java.net said:
This component actually makes sense;

don't forget the JS / Ajax stuff here ...

Also, perhaps we want to "polish" the <h:form> ?
=> Isn't doing "enctype" on component/tag a little bit lame?
Even what Trinidad has (usesUpload:boolean);
Perhaps the form should figure out if inputFile is in the game or not ... ?

-Matthias

**Comment by glassfishrobot** [comment]
mwessendorf@java.net said:
Toboago (the library is way better than its status) does that:
-It figures out if a upload widget is in the game and does the enctype stuff for
the (poor) page developer.

Trinidad makes it a little bit easier than the standard, with the usesUpload flag;

I personally would vote for the first, but if that is not desired by the EG,
introducing a convenience attribute (usesUpload) may make sense....

**Comment by glassfishrobot** [comment]
kenpaulsen said:
+1 for enabling this component to work via Ajax. While this adds a lot of
complexity to the implementation... it would be great if a page author could use
Ajax w/ this component the same way they use it with any other input component.

**Comment by glassfishrobot** [comment]
mojavelinux said:
> Perhaps the form should figure out if inputFile is in the game or not

Absolutely. The advantage of JSF is that we have the component tree in hand, so
let's use it. The form should definitely figure out by itself that the enctype
needs to be adjusted. This would be a huge win in the JSF column ![](https://java.net/jira/images/icons/emoticons/wink.gif)

And yes, we need to support Ajax as well. So this component has two separate
renderers, since the non-Ajax case is very simple whereas the Ajax case requires
more processing.

**Comment by glassfishrobot** [comment]
driscoll said:
There's a newish file upload spec, too, supported in FF 3.6\. We need to be
forward compatible.

[http://hacks.mozilla.org/2009/12/w3c-fileapi-in-firefox-3-6/](http://hacks.mozilla.org/2009/12/w3c-fileapi-in-firefox-3-6/)

**Comment by glassfishrobot** [comment]
ngriffin7a said:
While working on testing JSF 2.0 and the JSR 301 portlet bridge within Liferay
Portal, I came up with a simple way to support (Non-Ajax) file upload in JSF –
not only in a servlet environment, but in a portlet environment as well. It also
handles the problem in which the FaceletViewHandler restoreView(FacesContext
context, String viewId) method is unable to retrieve the javax.faces.ViewState
request parameter for HTML forms with enctype="multipart/form-data"

For more info, see:
[http://www.edorasframework.org/doc/1.3.0-SNAPSHOT/docbook/en-](http://www.edorasframework.org/doc/1.3.0-SNAPSHOT/docbook/en-)
US/html/pf-file-upload.html

Here is the proposal I was going to send to the EG just before 2.0 was released -

*   BTW it does not depend on Servlet 3.0.

PROPOSAL:

Support the following use case:

<h:form enctype="multipart/form-data">
...
</h:form>

===== BEGIN SPEC ADDITION TO:
javadocs/javax/faces/context/ExternalContext.html#getRequestParameterMap()

If the postback of the form is of enctype="multipart/form-data" then this
method is responsible for saving any file(s) that may be present in the request.
The files are saved on the server beneath a directory specified by the
"javax.faces.UPLOADED_FILES_DIR" context initialization parameter. If the
parameter is not present, then the "java.io.tmpdir" system property is consulted.
After all of the files are saved, this method sets a request attribute named
"javax.faces.UPLOADED_FILES" whose value is an immutable
java.util.Map<String, java.io.File>. The map key is the request parameter name
(clientId) and the map value shall be the uploaded java.io.File. It is the
responsibility of the application developer to delete the files from the filesystem
when they are no longer needed.

Servlet: In order to guarantee that multiple users can upload files with the same
filename, the files shall be saved in a sub-folder whose name is the value of
HttpSession.getId().

Portlet: In order to guarantee that multiple users can upload files with the same
filename, the files shall be saved in a sub-folder whose name is the value of
PortletSession.getId()

===== END SPEC ADDITION =====

**Comment by glassfishrobot** [comment]
mojavelinux said:
Seam took a different, arguably more complex, approach. A filter processes every
incoming request and if the method is POST and the content type starts with
"multipart/" it wraps the HttpServletRequest in a MultipartRequest object.

[http://anonsvn.jboss.org/repos/seam/branches/community/Seam_2_2/src/main/org/jboss/seam/web/MultipartFilter.java](http://anonsvn.jboss.org/repos/seam/branches/community/Seam_2_2/src/main/org/jboss/seam/web/MultipartFilter.java)
[http://anonsvn.jboss.org/repos/seam/branches/community/Seam_2_2/src/main/org/jboss/seam/web/MultipartRequest.java](http://anonsvn.jboss.org/repos/seam/branches/community/Seam_2_2/src/main/org/jboss/seam/web/MultipartRequest.java)
[http://anonsvn.jboss.org/repos/seam/branches/community/Seam_2_2/src/main/org/jboss/seam/web/MultipartRequestImpl.java](http://anonsvn.jboss.org/repos/seam/branches/community/Seam_2_2/src/main/org/jboss/seam/web/MultipartRequestImpl.java)

Then, in the decode method of the renderer for UIUpload, it reads the file data
from the MultipartRequest.

[http://anonsvn.jboss.org/repos/seam/branches/community/Seam_2_2/ui/src/main/java/org/jboss/seam/ui/renderkit/FileUploadRendererBase.java](http://anonsvn.jboss.org/repos/seam/branches/community/Seam_2_2/ui/src/main/java/org/jboss/seam/ui/renderkit/FileUploadRendererBase.java)

I think if we can make what Neil is suggesting work, it would be better than
having to introduce a filter. But perhaps some of the Seam code will prove helpful.

**Comment by glassfishrobot** [comment]
ngriffin7a said:
Servlet filters can be technology-incompatible with the portlet lifecycle.
Additionally, they require entries in web.xml

So I think in the interest of being compatible with portals, and to be zero-config
for the h:inputFile component, I think that modifying the ExternalContext.html#getRequestParameterMap() implementation is the easiest.

**Comment by glassfishrobot** [comment]
mojavelinux said:
+1

**Comment by glassfishrobot** [comment]
mojavelinux said:
Update target milestone to 2.1

**Comment by glassfishrobot** [comment]
@edburns said:
rogerk

**Comment by glassfishrobot** [comment]
rogerk said:
triage

**Comment by glassfishrobot** [comment]
rogerk said:
retarget

**Comment by glassfishrobot** [comment]
rogerk said:
triage

**Comment by glassfishrobot** [comment]
ngriffin7a said:
I've just implemented a solution very similar to my comment on December 11, 2009 in the PortletFaces Bridge. It's working quite nicely – no need for Servlet 3.0, no need for a filter. Everything works great. Making this work in a Servlet-based JSF like Mojarra should be a piece of cake.

**Comment by glassfishrobot** [comment]
@edburns said:
[http://java.net/jira/browse/JAVASERVERFACES_SPEC_PUBLIC-802](http://java.net/jira/browse/JAVASERVERFACES_SPEC_PUBLIC-802)

**Comment by glassfishrobot** [comment]
Marked as **duplicate** on Monday, January 31st 2011, 12:35:41 am

**Comment by glassfishrobot** [comment]
@manfredriem said:
Closing resolved issue out

**Comment by glassfishrobot** [comment]
This issue was imported from java.net JIRA JAVASERVERFACES_SPEC_PUBLIC-690
